pkg/agent/controller/egress/ip_scheduler.go

```diff
@@ -0,0 +1,286 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package egress
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/memberlist"
+	crdv1a2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlisters "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+)
+
+const (
+	// workItem is the only item that will be enqueued, used to trigger Egress IP scheduling.
+	workItem = "key"
+)
+
+// scheduleEventHandler is a callback when an Egress is rescheduled.
+type scheduleEventHandler func(egress string)
+
+// scheduleResult is the schedule result of an Egress, including the effective Egress IP and Node.
+type scheduleResult struct {
+	ip   string
+	node string
+}
+
+// egressIPScheduler is responsible for scheduling Egress IPs to appropriate Nodes according to the Node selector of the
+// IP pool, taking Node's capacity into consideration.
+// The scheduling process takes the spec of Egress and ExternalIPPool and the state of memberlist cluster as inputs,
+// generates scheduling results deterministically. It will be triggerred when any schedulable Egress changes or the
+// memberlist cluster changes.
+type egressIPScheduler struct {
+	// cluster is responsible for selecting a Node for a given IP and pool.
+	cluster memberlist.Interface
+
+	egressLister       crdlisters.EgressLister
+	egressListerSynced cache.InformerSynced
+
+	// queue is used to trigger scheduling. Triggering multiple times before the item is consumed will only cause one
+	// execution of scheduling.
+	queue workqueue.Interface
+
+	// mutex is used to protect scheduleResults.
```

Comment:
We have a single worker today. Do we still need the mutex?

Reply:
The results could be read by workers in egress controllers via `GetEgressIPAndNode` concurrently.

Reply:
Should we use RW mutex if there are multiple readers?

Reply:
Its type is indeed RWMutex